File: Sources/lambda/Services/TopTracksService.swift
import Foundation
import Logging

actor TopTracksService {
    private let spotifyApi: SpotifyApi
    private let logger: Logger

    init(spotifyApi: SpotifyApi, logger: Logger) {
        self.spotifyApi = spotifyApi
        self.logger = logger
    }

    func handleTopTracks(timeRange: String, limit: Int) async throws -> TopTracksApiResponse {
        do {
            let response = try await spotifyApi.getTopTracks(timeRange: timeRange, limit: limit)

            let tracks = response.items.map { item in
                TopTrackResponseItem(
                    title: item.name,
                    artist: item.artists.map(\.name).joined(separator: ", "),
                    album: item.album.name,
                    albumImageUrl: item.album.images.first?.url ?? "",
                    songUrl: item.externalUrls.spotify
                )
            }

            return TopTracksApiResponse(tracks: tracks)
        } catch {
            if let spotifyError = error as? SpotifyServiceError {
                logger.error("Top tracks fetch failed: \(spotifyError.description)")
            } else {
                logger.error("Top tracks fetch failed: \(error.localizedDescription)")
            }
            throw error
        }
    }
}

File: src/Lho.Lambda/Services/TopTracksService.cs
using Amazon.Lambda.Core;
using Lho.Lambda.Clients.Spotify;
using Lho.Lambda.Models;
using Lho.Lambda.Observability;

namespace Lho.Lambda.Services;

public class TopTracksService(SpotifyApi spotifyApi, ILambdaLogger logger)
{
  public async Task<TopTracksApiResponse> HandleTopTracks(string timeRange, int limit)
  {
    try
    {
      var response = await spotifyApi.GetTopTracks(timeRange, limit);
      var tracks = response.Items
        .Select(item => new TopTrackResponseItem(
          Title: item.Name,
          Artist: string.Join(", ", item.Artists.Select(artist => artist.Name)),
          Album: item.Album.Name,
          AlbumImageUrl: item.Album.Images.FirstOrDefault()?.Url ?? "",
          SongUrl: item.ExternalUrls.Spotify))
        .ToArray();

      return new TopTracksApiResponse(tracks);
    }
    catch (Exception exception)
    {
      logger.LogLine($"Top tracks fetch failed: {exception}");
      await SentryTelemetry.CaptureExceptionAsync(exception, logger, new Dictionary<string, string?>
      {
        ["operation"] = "top-tracks",
        ["time_range"] = timeRange
      });
      throw;
    }
  }
}
